go/database/mpt/verification_test.go

```diff
@@ -412,6 +414,118 @@ func TestVerification_ValueNodeHashModificationIsDetected(t *testing.T) {
 		}
 	})
 }
+
+func TestVerification_MissingCodeHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		if err := VerifyCodesAndForest(dir, config, roots, NilVerificationObserver{}); err == nil {
+			t.Errorf("missing hash in code file should have been detected")
+		}
+	})
+}
+
+func TestVerification_DifferentHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		testHash := common.Keccak256([]byte{1})
+		codes := map[common.Hash][]byte{
+			testHash: {2},
+		}
+		if err := writeCodes(codes, filepath.Join(dir, "codes.dat")); err != nil {
+			t.Fatalf("failed to write code file")
+		}
+
+		if err := VerifyCodesAndForest(dir, config, roots, NilVerificationObserver{}); err == nil {
+			t.Errorf("different hash in code file should have been detected")
+		}
+	})
+}
+
+func TestVerification_ExtraCodeHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		ctrl := gomock.NewController(t)
+		observer := NewMockVerificationObserver(ctrl)
+
+		testHash1 := common.Keccak256([]byte{1})
+		testHash2 := common.Keccak256([]byte{2})
+		codes := map[common.Hash][]byte{
+			testHash1: {1},
+			testHash2: {2},
+		}
+
+		// TODO - this is fragile, it depends on precise strings
```

Comment:
I thought you resolve this todo :)

I think it would be better to let the observer pass through all strings and just make sure certain substring was passed.